src/librustc_typeck/check/coercion.rs

```diff
@@ -121,39 +131,107 @@ impl<'f, 'tcx> Coerce<'f, 'tcx> {
         // See above for details.
         match b.sty {
             ty::TyRawPtr(mt_b) => {
-                return self.unpack_actual_value(a, |a| {
-                    self.coerce_unsafe_ptr(a, b, mt_b.mutbl)
-                });
+                return self.coerce_unsafe_ptr(a, b, mt_b.mutbl);
             }
 
             ty::TyRef(_, mt_b) => {
-                return self.unpack_actual_value(a, |a| {
-                    self.coerce_borrowed_pointer(expr_a, a, b, mt_b.mutbl)
-                });
+                return self.coerce_borrowed_pointer(expr_a, a, b, mt_b.mutbl);
             }
 
             _ => {}
         }
 
-        self.unpack_actual_value(a, |a| {
-            match a.sty {
-                ty::TyBareFn(Some(_), a_f) => {
-                    // Function items are coercible to any closure
-                    // type; function pointers are not (that would
-                    // require double indirection).
-                    self.coerce_from_fn_item(a, a_f, b)
-                }
-                ty::TyBareFn(None, a_f) => {
-                    // We permit coercion of fn pointers to drop the
-                    // unsafe qualifier.
-                    self.coerce_from_fn_pointer(a, a_f, b)
-                }
-                _ => {
-                    // Otherwise, just use subtyping rules.
-                    self.subtype(a, b)
+        let b = self.fcx.infcx().shallow_resolve(b);
+        match a.sty {
+            ty::TyBareFn(Some(_), a_f) => {
+                // Function items are coercible to any closure
+                // type; function pointers are not (that would
+                // require double indirection).
+                return self.coerce_from_fn_item(a, a_f, b);
+            }
+            ty::TyBareFn(None, a_f) => {
+                // We permit coercion of fn pointers to drop the
+                // unsafe qualifier.
+                return self.coerce_from_fn_pointer(a, a_f, b);
+            }
+            _ => {}
+        }
+
```

Comment:
so, in the code before this PR, the control flow that reached this point (line 159) corresponds to the comment above "Otherwise, just use subtyping rules."

Clearly you are no longer "just using the subtyping rules" in the code that follows that point in the control flow here.  :)

Still, for some reason I wish there was a comment line or two here of the same pithy nature as "Otherwise just use subtyping rules" before jumping straight into the discussion of "generalization in hopes of an unsizing coercion" ... something that helped the reader re-establish what case we are in in the coercion analysis?  Or is that just a pipe dream?